include/npcomp/Dialect/TCP/IR/TCPOps.td

```diff
@@ -42,6 +42,29 @@ def TCP_MaxOp : BinaryArithmeticOp<"max"> {
   }];
 }
 
+class UnaryArithmeticOp<string mnemonic, list<OpTrait> traits = []> :
```

Comment:
I don't have a real opinion on this yet, but a part of me is wondering whether we want every individual op in the TCF dialect versus having some kind of way to just represent a `tcf.unary_arithmetic "abs"` or have a region where we embed the TCP op. It seems to me that at the TCF level, the type of op more has to do with how it interacts with the dataflow/types/shapes vs what it does.

Reply:
Yeah, I've been thinking about this too. There's a continuum here of elementwise ops that can go all the way from tcf to linalg . The "scalar body" is the same, but the control flow (possibly even TCF binary broadcasting and unranked loop structure) changes around it. (modulo TCF wanting dtype-agnostic; we could do something like your ufunc overloads)

I'm just trying to be lazy about designing anything until we actually have an e2e use case.

I wouldn't be surprised if most flows don't want *any* algebraic transforms at TCF level, and focus more on shape/type refinement stuff and importing/exporting to adjacent formats. That would push us more toward a "I'd never want to do `dyn_cast<tcf::ExpOp>`" situation which would make a tcf.unary_elementwise more appealing.

Also, the current state looks a lot more uniform than things really are for the elementwise landscape, making it unclear if this generalizes beyond tcf.unary_elementwise with same operand+result type. Examples of complications:
- differing dtypes on the arguments. E.g. the predicate on `select`, the src/dst types on sitofp and fptosi
- extra attributes that only exist on some elementwise ops, e.g. the "eq" in  `cmp "eq"`
- different traits on each op, e.g. sub is not commutative but add is.
- some ops only make sense on integers, e.g. xor, "logical not", "shift right arithmetic"

We'd probably want more sophisticated verification and using mlir's native trait system for these cases. Given this, I'm still not at the point that I'm willing to fragment off what I think is a pretty small subset of the unary ops. Maybe something like `tcf.special_math_function "tanh"` would be a natural fit (but that still doesn't even cover special math functions super well; e.g. sperical bessel functions are binary operators).